### PR TITLE
Remove hacky code added in 2482 and feature enhancement

### PR DESCRIPTION
**Pull request by sklam**
(base on #2482)

* add runtime check to ensure the compiler didn't incorrectly assumed constantness
    * ~~~TODO: check if llvm will remove the check~~~ (Yes, it does)
* refactor the way an impl indicates it can accept `types.Const` to not use exception as a signal.
* make np.sum specialize to constant literals for axis without limit.
    * dynamic axis value is still limited to 0, 1, 2, 3
* support axis as kwargs e.g. `np.sum(axis=1)`

### CODE REVIEW

**Comment by seibert** [comment]
OK, you should be able to re-merge master and shorten the commit list now.

File: numba/tests/test_array_methods.py [hunk 1]
```diff
@@ -630,13 +642,32 @@ def test_sum(self):
...
+        a = np.ones((7, 6, 5, 4, 3))
+        self.assertPreciseEqual(pyfunc(a, axis=1), cfunc(a, axis=1))
+        # OK
+        self.assertPreciseEqual(pyfunc(a, axis=2), cfunc(a, axis=2))
+
+    def test_sum_const(self):
+        pyfunc = array_sum_const_multi
+        cfunc = jit(nopython=True)(pyfunc)
+
+        arr = np.ones((3, 4, 5, 6, 7, 8))
```
Comment (stuartarchibald) [inline numba/tests/test_array_methods.py]: Perhaps check a negative const axis too ?

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -198,21 +211,27 @@ def array_sum_impl_axis(arr, axis):
...
-                index_tuple2 = _gen_index_tuple(arr.shape, axis_index, 1)
-                result += arr[index_tuple2]
-            elif axis == 2:
-                index_tuple3 = _gen_index_tuple(arr.shape, axis_index, 2)
-                result += arr[index_tuple3]
-            elif axis == 3:
-                index_tuple4 = _gen_index_tuple(arr.shape, axis_index, 3)
-                result += arr[index_tuple4]
+            if is_axis_const:
+                # constant specialized version works for any valid axis value
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: I think the docs probably need updating to reflect this change.

File: numba/tests/test_array_methods.py (hunk 1 above)
Comment (sklam) [inline numba/tests/test_array_methods.py]: Glad you mention this.  Cause I forgotten to test this and a small addition will enable the support of negative axis.

Comment (sklam) [inline numba/tests/test_array_methods.py]: Negative values are now handled and tested.

File: numba/targets/arraymath.py (hunk 2 above)
Comment (sklam) [inline numba/targets/arraymath.py]: Docs updated

**Comment by stuartarchibald** [comment]
Thanks for the fixes (especially 93ebcce). This seems good on inspection, can be merged once CI passes.

**Pull request merged**